src/matches.js

```diff
@@ -2,12 +2,16 @@ function fuzzyMatches(
   textToMatch,
   node,
   matcher,
-  {collapseWhitespace = true, trim = true} = {},
+  {collapseWhitespace = true, trim = true, normalizer} = {},
```

Comment:
I would prefer it if we could accomplish `collapseWhitespace` and `trim` from within a built-in `normalizer`. That normalizer can take these options into account:

```js
{collapseWhitespace = true, trim = true, normalizer = getDefaultNormalizer({collapseWhitespace, trim})} = {}
```

It may also be a good idea to soft-deprecate `collapseWhitespace` and `trim` in favor of providing a custom normalizer. This would amount to removing the docs for those options and replacing it with docs for `normalizer`. Then in a future version we can remove the options and have it built-into our default normalizer.

I just don't like having multiple ways to do the same thing, and I like the normalizer as it's more powerful.

Reply:
Yep, I definitely like the idea of removing `collapseWhitespace` and `trim`, but I didn't want to be the one to suggest it! :)

Just to check my understanding, let me rephrase what I think you're proposing:

a) publicly expose a new thing called `getDefaultNormalizer` that takes `collapseWhitespace` and `trim` options
b) soft-deprecate `collapseWhitespace` and `trim` by not documenting them any more
c) the default value of the `normalizer` option becomes the output of `getDefaultNormalizer`, called with whatever `trim` and `collapseWhitespace` options are provided (until we deprecate + remove them)

I'm not a huge fan of the idea of an array of normalizers: functions are nicely composable already, and with an array you get issues about who owns the array, how shared is it, do I have to clone it before pushing my normalizer, etc?  If you just have one function, composition is straightforward to comprehend, and you don't need to reimplement anything: you just compose over the default `getDefaultNormalizer` if you want it.

(In fact, my team have already doubled up all the queries, producing ...WithoutUCC versions, such as `getByTextWithoutUCC`, and they still accept a `normalizer` but chain it with a `withoutUCC` normalizer, which they choose to run first.)

I'll update the PR.

Reply:
Hmm. One little thought:  let's say we have a user who writes
`queryBySomething(..., { trim: true, collapseWhitespace: false, normalizer: customNormalizer })`.

Long term we're obviously suggesting that `trim` and `collapseWhitespace` would be removed, but in the short term what should that call do where they specify values for `trim` and `collapseWhitespace` _and_ `normalizer` (which is supposed to replace both)?

Should we error, saying that they can't specify a custom `normalizer` _and_ values for `trim`/`collapseWhitespace`?  I suspect we _should_ error, otherwise we're temporarily giving them an API that will stop working later when we remove the properties, whereas if we error, we force them to do the right thing - composing with `getDefaultNormalizer` from the get go - and that'll continue to work even when `trim` + `collapseWhitespace` are removed as first-class parameters...?

Reply:
Hmm, one more idea: With an _object_ you could turn off a default normalizer by passing false or null instead of a function, so the API is almost the same as the current top-level options, and makes it easier to not accidentally opt out of new defaults. Depending on implementation, you could probably also accept an array of (optionally anonymous) functions, and may as well accept a single function as well.

```
normalizers/
  trim.js
  collapseWhitespace.js
  index.js (exports all)
```

```js
import * as normalizers from 'dtl/(dist)/normalizers'
// Or
import normalizers from 'dtl'

getByText(document.body, 'abc', {
  normalizers,
})

// Accept a single function as well, that doesn't require any option handling
getByText(document.body, 'abc', someCustomFn)

getByText(document.body, 'abc', {
  normalizers: {
    // Possibly merge by default if receiving an object?
    ...normalizers,
    // Falsy values don't get called
    trim: false,
    collapseWhitespace: null,
    // A custom function
    replaceUnicode,
  }
})```

Reply:
I think discovery would be a little harder with that API: firstly you've got two different things called `normalizers`: an import, which I'm guessing is a _default_ set of normalizers, and a parameter specifying a chosen bunch of normalizers.

Ordering also isn't very obvious: if I were to specify `trim: true` and a custom normalizer, which would run first? Or two custom normalizers? I can't tell or control it.  Having an artbitrary-extensible object also mixed in with well-known named parameters is a bit tricky: you wouldn't ever be able to add more well-known named parameters later for fear of confusion with user-specified ones, and mixing booleans and functions in a single object is a little unusual.

I still like @kentcdodds's suggestion of specifying a single `normalizer` function with a `getDefaultNormalizer` function exposed. It feels really obvious what's going on, and both user and library extensibility is simple and separate.

Reply:
Sorry, this comment is a bit long!

tl;dr:
- probably need a config option to set the default
- per-test overrides given you have a custom setup on top of the default are a bit complex to do right

Assumptions:

- Usually you want a suite-wide default plus a way of turning _off_ very specific things for certain tests to check for edge cases. 
- You generally don't want to "assume full control of normalization"; you want to extend the default behavior in a forwards-compatible way so you get new features and bug fixes, while adding in support for some specific additional cases.

Playing out the use cases using the proposed API:

1. If you're writing something you're going to use across the whole app, you'll probably want some way to set the default normalizer function so you don't have to add normalizer options to every single query. You'll also probably want some way of customizing the options, so you'll need an internal version of `getDefaultNormalizer` which may even wrap that function and all its options (doing something to avoid namespace collisions).

---

2. If you're doing a one-off normalizer override:

```js
// ignore defaults completely
// 🚀boldly taking normalization into our own hands
getByText(elem, 'text', {
  normalizer: customNormalizer,
})

// take options for custom factory, but ignore forwards compat completely
// (don't call default normalizer even if options match)
// 🤔 what if we want to track upstream functionality?
getByText(elem, 'text', {
  normalizer: getCustomNormalizer({
    trim: false, // don't care if this is a default option
    removeFoo: false,
  })
})

// with default handler and default options
// assuming customNormalizer takes and returns a string
// 👍 composition ftw
getByText(elem, 'text', {
  normalizer: str => 
    customNormalizer(defaultNormalizer(str)),
})

// with default handler and default options
// assuming customNormalizer takes and returns a string
// a realistic use case if you have any conditional logic
// or multiple custom normalization functions
// still a 👍, but you probably wouldn't want to inline this
getByText(elem, 'text', {
  normalizer: str => {
	const defaultResult = defaultNormalizer(str)
    // ...some custom logic here...
    const evenMoreNormalized = customNormalizer(defaultResult)
    return evenMoreNormalized
  },
})

// using the getter to customize default normalizer
// assuming the customNormalizer takes and returns a string
// 🤔 a bit complicated now
getByText(elem, 'text', {
  normalizer: str => 
    customNormalizer(
      getDefaultNormalizer({ trim: false})(str)
    ),
})

// using factories for both default and custom
// custom has no options but takes a function and returns a string
// 🤔 potentially fragile, what if we want options later?
getByText(elem, 'text', {
  normalizer: getCustomNormalizer(getDefaultNormalizer({ trim: false})),
})

// using factories for both default and custom
// custom has options as the second argument
// 🤔 not really leveraging composition at this point
getByText(elem, 'text', {
  normalizer: getCustomNormalizer(
    getDefaultNormalizer({ trim: false}),
    {removeUCC: false}
  ),
})

// hide default normalizer inside custom normalizer
// add default options to custom normalizer factory
// 🤔 potential name collisions
getByText(elem, 'text', {
  normalizer: getCustomNormalizer({
    removeUCC: false,
    trim: false
  }),
})

// hide default normalizer inside custom normalizer
// take default options as separate argument
// 🤔not that composable any more
getByText(elem, 'text', {
  normalizer: getCustomNormalizer(
    { removeUCC: false },
    {trim: false}
  ),
})

// add default options to custom normalizer factory under a namespace
// 🤔 not that composable any more
getByText(elem, 'text', {
  normalizer: getCustomNormalizer({
    removeUCC: false,
    defaultNormalizerOptions: {
      trim: false,
    },
  }),
})
```
---

3. Defining a one-off handler assuming you have set a global custom handler:

(All patterns from above apply except now the default behavior is split between the library default and the custom default)

Reply:
That feels... complicated. I'd love to try and keep the library simple...

Reply:
Yes, that's rather a complex API, plus I (as somebody who actually needs to use a custom normalizer) don't want to apply the same normalizer everywhere.

I'm just finishing up the docs changes for the API I posted at https://github.com/kentcdodds/dom-testing-library/pull/172#discussion_r240053817. That API feels pretty good to use in practice.

Reply:
The API in the code samples I posted is the one in your comment, just fleshed out with more scenarios. I think it looks OK for one-offs except some complexity when inlining the getter that returns a function. If you want to use this project-wide it does seem a bit difficult especially if you want to add options to the custom function or compose several conditionally.

Reply:
I'm ok with going with the proposed API and dealing with any global config or alternate APIs later if someone asks for them, unless someone has a better idea to address those right now.